### ISSUE TITLE
Viewport selection management

### ISSUE THREAD

**Issue by ronen** [rprt/rspec-page-regression#25]
It would be good to add the following features:

``` ruby
config.default_viewports = [:small, :medium]   # default is all defined viewports

match_reference_screenshot viewport: :small
match_reference_screenshot viewport: [:small, :medium]
match_reference_screenshot except_viewport: :large
match_reference_screenshot except_viewport: [:small, :large]
```

and they should all check that the arguments actually match defined viewport  names.

**Comment by paresharma** [comment]
Yeah, this has always been under consideration. I agree that this will become quite a necessary option once we release the multiple viewports feature. We should really press for this to be part of the v1.0 release.

But there is one thing, in my opinion when someone( or I) define a config with multiple viewports I want to test my web pages against these viewport sizes. This is already the default. So, to introduce the `default_viewports` config in between this and the option to choose or except viewport size(s) at the matcher level feels a bit redundant.

What is your opinion on this? You do have the option to totally convince me though. :smile:

**Comment by ronen** [comment]
> Yeah, this has always been under consideration

Yeah, just wanted to get an issue listed so we don't forget about it & have somewhere to discuss it.

> So, to introduce the default_viewports config in between this and the option to choose or except viewport size(s) at the matcher level feels a bit redundant.

Here are some scenarios I can imagine.  Probably the second one is more compelling:
- Suppose your app doesn't display properly on tiny screens such as iPhone3, instead displays a "sorry your screen is too small" page.  You define multiple viewports, says `:tiny` `:mobile`, `:tablet`, `:desktop`.  You want the default behavior to be `config.default_viewports = [:mobile, :tablet, :deskop]` and ignore `:tiny`, but there'll be one test for the "sorry" page for which you'll specify `match_reference_screenshot viewport: :tiny`
- You defined multiple viewports, say `:mobile`, `:tablet`, and `:desktop`--but doing all those screenshots is making your tests run very slowly!  Since you're designing mobile first using Bootstrap 3 or Foundation 5, the majority of pages just scale up automatically and you don't need to explicitly test all sizes.  So you specify `config.default_viewports = [:mobile]` and now your tests run 3 times as fast--yay!  Only where you've needed to tweak the layout to handle other screen sizes differently do you specify `match_reference_screenshot viewport: [:mobile, :desktop]` or `match_reference_screenshot viewport: [:mobile, :tablet, :desktop]` to check at multiple sizes.

**Comment by paresharma** [comment]
Ahh, I am totally convinced and the second pointer is indeed compelling. Also, I think we already have a very good wiki material this feature :smile:. I am gonna start working on this as the next PR.

**Comment by paresharma** [comment]
I think, now that we have introduced multiple viewports, there is a possible gap in the functionality where we allow custom `reference_screenshot_path`.

```
 # spec_helper.rb:
 RSpec::PageRegression.configure do |config|
   config.viewports = {
                        desktop: [1280, 1024],
                        tablet: [1024, 768],
                        mobile: [480, 320]
                      }
 end

# in a spec file
expect(page).to match_reference_screenshot "/path/to/my/screenshot.png"
```

Now we're expecting one of the pages to be the same on different viewport sizes. I think we should allow overriding the reference_screenshot_path for one viewport only. Something like:

```
expect(page).to match_reference_screenshot reference_screenshot_path: "/path/to/my/screenshot.png", viewport: :mobile
```

Of course, this `:viewport` will be implicit if there is only one viewport configured.

I thinking of introducing this along-with this feature.

@ronen Any thoughts on this?

**Comment by ronen** [comment]
@paresharma good catch!

Here's my thoughts:
- In any case I think the path should be a keyword arg (as you propose), not a positional arg (as it was pre 1.0)
- I'd call the arg just `path:`, I think that should be unambiguous
- I agree that if there's only one viewport (due to config or matcher options), then `path:` can be given as just a single value, as per your example.
- Otherwise I'd say why not allow a hash `path: { mobile: "/path/to/mobile/screenshot.png", tablet: "/path/to/tablet/screenshot.png", etc. }`

**BUT**...

Is this option actually useful?  Should it just be retired for 1.0 (until/unless somebody asks for it)?  I put it in initially for completeness but never had any use for it myself.  I can't actually think of a use case where somebody would locate the reference image(s) for one example in an oddball location.  More useful I'd guess would be to be able to globally configure the directory where the reference screenshots should be rooted, in case `specs/reference_screenshots` doesn't fit the bill.

@paresharma, @jekuno  Any thoughts on this?

**Comment by paresharma** [comment]
Actually I have never really used this option. I am all for retiring it for now and see how it goes. I like the idea of providing the option to configure it globally though.
@ronen @jekuno Should we go ahead with these changes?

**Comment by jekuno** [comment]
Sometimes I need two reference screenshots per test e.g. when testing both the initial and the final state (e.g. before and after clicking a button). That's where I need to distinguish the separate screenshots by name. However it would be sufficient to provide a base filename (without path and file extension) using a `filename` option. I would prefer this over specifying a complete path.

**Comment by ronen** [comment]
OK, seems we're all agreed to retire this option.  Great, less code!

BTW the config option is already an issue, #18  :)

> However it would be sufficient to provide a base filename (without path and file extension) using a `filename` option. I would prefer this over specifying a complete path.

@jekuno I'd  suggest the option be something like `label:` rather than `filename:`, since it'd be just one more component added to the mix of path, viewport, and extention--it's identifying the screenshot conceptually rather than actually specifying a file.

**Comment by jekuno** [comment]
@ronen Agree. `label` or `basename` fits better than `filename`.

**Comment by ronen** [comment]
I added #26

**Comment by jekuno** [comment]
:+1:
